### PR TITLE
One more bunch of fixes for oss-fuzz

### PR DESCRIPTION
**Pull request by Jakuje**
##### Checklist

- [ ] PKCS#11 module is tested
- [ ] Windows minidriver is tested
- [ ] macOS tokend is tested

### CODE REVIEW

File: src/libopensc/card-coolkey.c [hunk 1]
```diff
@@ -1757,6 +1757,9 @@ static int coolkey_rsa_op(sc_card_t *card,
 		u8 len_buf[2];
 		size_t out_length;
 
+		/* Free card response -- nothing useful -- result is in export object */
+		free(*crypt_out_p);
```
Comment (frankmorgner) [inline src/libopensc/card-coolkey.c:1761]: Are you sure that this is correct? It looks like there is some case where this calls free on a buffer on the heap:
https://github.com/OpenSC/OpenSC/blob/0d958c1f7d8ea34e3e127bf560dcbf963b9147a9/src/libopensc/card-coolkey.c#L1702-L1703

Comment (Jakuje) [reply src/libopensc/card-coolkey.c:1761]: The code is very confusing, but most important is that `crypt_in_len = datalen;` and therefore the conditions `if (crypt_in_len > MAX_COMPUTE_BUF) {` and `if (datalen > MAX_COMPUTE_BUF) {` are equal. In the first one we are setting

https://github.com/OpenSC/OpenSC/blob/0d958c1f7d8ea34e3e127bf560dcbf963b9147a9/src/libopensc/card-coolkey.c#L1725

which keeps the io function to allocate the memory. This probably does not happen with real cards, but fuzzer can return some mess somethimes.

Comment (frankmorgner) [reply src/libopensc/card-coolkey.c:1761]: `crypt_out_p` is only reset `if (crypt_in_len > MAX_COMPUTE_BUF)`, so there seems to be the other case where it is *not* set to `NULL`. Would it be possible to clearify the code, e.g. by using different variables?

Comment (Jakuje) [reply src/libopensc/card-coolkey.c:1761]: I will try to improve that part.

Comment (Jakuje) [reply src/libopensc/card-coolkey.c:1758]: Rewritten in a8101c26. Does it look better for you now?

Comment (Jakuje) [reply src/libopensc/card-coolkey.c:1759]: The final version is in 6b66c43 as the initial had some issues to build. Looks fine now.

**Comment by Jakuje** [comment]
I added one more report from oss-fuzz today and rebased on top of current master.

File: src/libopensc/pkcs15-sc-hsm.c [hunk 2]
```diff
@@ -862,6 +864,7 @@ static int sc_pkcs15emu_sc_hsm_init (sc_pkcs15_card_t * p15card)
 	appinfo->aid = sc_hsm_aid;
 
 	appinfo->ddo.aid = sc_hsm_aid;
+	sc_pkcs15_free_app(p15card);
```
Comment (frankmorgner) [inline src/libopensc/pkcs15-sc-hsm.c:867]: I assume some other initialization routine initialized this, but failed in the end (`sc_pkcs15_bind_internal()`?) This routine would be a better place for cleaning this up, wouldn't it?

**Comment by frankmorgner** [comment]
(Big thank you, for keeping up with the reported issues :tulip: )

File: src/libopensc/pkcs15-sc-hsm.c (hunk 2 above)
Comment (Jakuje) [reply src/libopensc/pkcs15-sc-hsm.c:867]: Actually, this one was more shot in the dark. Correct fix was 31e442f3 as you propose in pkcs15. I will drop ddea3c73 and 7d40797 as they do not seem to be much helpful.

**Pull request merged**